src/codegen/codegen_acc_visitor.cpp

```diff
@@ -265,53 +275,59 @@ std::string CodegenAccVisitor::get_variable_device_pointer(const std::string& va
     if (info.artificial_cell) {
         return variable;
     }
-    return "reinterpret_cast<{}>( nt->compute_gpu ? acc_deviceptr({}) : {} )"_format(type,
-                                                                                     variable,
-                                                                                     variable);
+    return "nt->compute_gpu ? cnrn_target_deviceptr({}) : {}"_format(variable, variable);
 }
 
 
 void CodegenAccVisitor::print_newtonspace_transfer_to_device() const {
     int list_num = info.derivimplicit_list_num;
-    printer->add_line("if (nt->compute_gpu) {");
-    printer->add_line("    auto device_vec = static_cast<double*>(acc_copyin(vec, vec_size));");
-    printer->add_line("    auto device_ns = static_cast<NewtonSpace*>(acc_deviceptr(*ns));");
-    printer->add_line("    auto device_thread = static_cast<ThreadDatum*>(acc_deviceptr(thread));");
+    printer->start_block("if(nt->compute_gpu)");
+    printer->add_line("double* device_vec = cnrn_target_copyin(vec, vec_size / sizeof(double));");
+    printer->add_line("void* device_ns = cnrn_target_deviceptr(*ns);");
+    printer->add_line("ThreadDatum* device_thread = cnrn_target_deviceptr(thread);");
     printer->add_line(
-        "    acc_memcpy_to_device(&(device_thread[{}]._pvoid), &device_ns, sizeof(void*));"_format(
+        "cnrn_target_memcpy_to_device(&(device_thread[{}]._pvoid), &device_ns);"_format(
```

Comment:
just curious - was auto deduced type was different here?

Reply:
I think for this one it was important to have `device_ns` be `void*` rather than `NewtonSpace*`.

This is because now we have
```c++
template <typename T>
void cnrn_target_memcpy_to_device(T*, T*) { ... }
```
and if you pass `void**` and `NewtonSpace**` it will fail to deduce `T`. We have to pass two pointers of the same type (up to const qualification).